File: core/common/src/main/kotlin/br/com/firstsoft/core/common/hardwaremonitor/HardwareMonitorData.kt
package br.com.firstsoft.core.common.hardwaremonitor

import br.com.firstsoft.core.common.hwinfo.SensorReadingType
import kotlinx.serialization.Serializable

@Serializable
data class HardwareMonitorData(
    val LastPollTime: Long,
    val Hardwares: List<Hardware>,
    val Sensors: List<Sensor>
) {
    @Serializable
    data class Hardware(
        val Name: String,
        val Identifier: String,
        val HardwareType: HardwareType
    )

    @Serializable
    data class Sensor(
        val Name: String,
        val Identifier: String,
        val HardwareIdentifier: String,
        val SensorType: SensorType,
        val Value: Float
    )

    enum class HardwareType {
        Motherboard,
        SuperIO,
        Cpu,
        Memory,
        GpuNvidia,
        GpuAmd,
        GpuIntel,
        Storage,
        Network,
        Cooler,
        EmbeddedController,
        Psu,
        Battery;
    }

    enum class SensorType {
        Voltage,
        Current,
        Power,
        Clock,
        Temperature,
        Load,
        Frequency,
        Fan,
        Flow,
        Control,
        Level,
        Factor,
        Data,
        SmallData,
        Throughput,
        TimeSpan,
        Energy,
        Noise;
    }
}

fun HardwareMonitorData.gpuReadings() = readings("GPU")
fun HardwareMonitorData.cpuReadings() = readings("CPU")
fun HardwareMonitorData.networkReadings() = readings("/nic/")
fun HardwareMonitorData.readings(namePart: String): List<HardwareMonitorData.Sensor> {
    return Sensors.filter { it.Identifier.contains(namePart, true) || it.Name.contains(namePart, true) }
        .sortedBy { it.SensorType }
}
fun HardwareMonitorData.getReading(identifier: String) = Sensors.firstOrNull { it.Identifier == identifier }

val HardwareMonitorData.FPS: Int
    get() = (getReading("/presentmon/presented")?.Value?.toInt() ?: 0).coerceAtMost(480)

val HardwareMonitorData.Frametime: Float
    get() = (getReading("/presentmon/frametime")?.Value ?: 0f).coerceAtLeast(0f).coerceAtMost(99f)

val HardwareMonitorData.RamUsage: Float
    get() = Sensors.firstOrNull { it.Name == "Memory Used" }?.Value?.coerceAtLeast(1f) ?: 1f

val HardwareMonitorData.RamUsagePercent: Float
    get() = RamUsage / (RamUsage + (Sensors.firstOrNull { it.Name == "Memory Available" }?.Value?.coerceAtLeast(1f) ?: 1f))

File: app-tauri/src/lib/model/hardwareMonitorData.ts
// Mirror of app/core/common/.../hardwaremonitor/HardwareMonitorData.kt

export enum HardwareType {
  Motherboard = 0,
  SuperIO = 1,
  Cpu = 2,
  Memory = 3,
  GpuNvidia = 4,
  GpuAmd = 5,
  GpuIntel = 6,
  Storage = 7,
  Network = 8,
  Cooler = 9,
  EmbeddedController = 10,
  Psu = 11,
  Battery = 12,
  Unknown = 13,
}

// HardwareMonitorData.SensorType in Kotlin — renamed to avoid clashing with the
// UI-level SensorType enum used by the settings.
export enum HwSensorType {
  Voltage = 0,
  Current = 1,
  Power = 2,
  Clock = 3,
  Temperature = 4,
  Load = 5,
  Frequency = 6,
  Fan = 7,
  Flow = 8,
  Control = 9,
  Level = 10,
  Factor = 11,
  Data = 12,
  SmallData = 13,
  Throughput = 14,
  TimeSpan = 15,
  Energy = 16,
  Noise = 17,
  Unknown = 18,
}

export interface HwHardware {
  Name: string
  Identifier: string
  HardwareType: HardwareType
}

export interface HwSensor {
  Name: string
  Identifier: string
  HardwareIdentifier: string
  SensorType: HwSensorType
  Value: number
}

export interface HardwareMonitorData {
  LastPollTime: number
  Hardwares: HwHardware[]
  Sensors: HwSensor[]
  PresentMonApps: string[]
}

// --- helpers (mirror the Kotlin extension functions) ---

export function readings(data: HardwareMonitorData, namePart: string): HwSensor[] {
  const part = namePart.toLowerCase()
  return data.Sensors.filter(
    (s) =>
      s.Identifier.toLowerCase().includes(part) ||
      s.Name.toLowerCase().includes(part),
  ).sort((a, b) => a.SensorType - b.SensorType)
}

export const gpuReadings = (data: HardwareMonitorData) => readings(data, "GPU")
export const cpuReadings = (data: HardwareMonitorData) => readings(data, "CPU")
export const networkReadings = (data: HardwareMonitorData) => readings(data, "/nic/")

export function getReading(
  data: HardwareMonitorData,
  identifier: string,
): HwSensor | undefined {
  return data.Sensors.find((s) => s.Identifier === identifier)
}

export function fps(data: HardwareMonitorData): number {
  const frametime = getReading(data, "/presentmon/frametime")?.Value ?? 1
  return Math.trunc(1000 / frametime)
}

export function frametime(data: HardwareMonitorData): number {
  return Math.max(0, getReading(data, "/presentmon/frametime")?.Value ?? 0)
}

export function ramUsage(data: HardwareMonitorData): number {
  return Math.max(
    1,
    data.Sensors.find((s) => s.Name === "Memory Used")?.Value ?? 1,
  )
}

export function ramUsagePercent(data: HardwareMonitorData): number {
  const used = ramUsage(data)
  const available = Math.max(
    1,
    data.Sensors.find((s) => s.Name === "Memory Available")?.Value ?? 1,
  )
  return used / (used + available)
}
